File: week09/20/pr5_Kth_smallest_Element_in_a_BST.c

//   Definition for a binary tree node.
struct TreeNode {
      int val;
      struct TreeNode *left;
      struct TreeNode *right;
};
 
void helper(struct TreeNode* node, int* cnt, int* result){
    if(!node) return;

    helper(node->left,cnt,result);

    (*cnt)--;
    if(*cnt==0){
        *result = node->val;
       return;
    }
    helper(node->right,cnt,result);

}

int kthSmallest(struct TreeNode* root, int k) {
    int result;
    helper(root,&k,&result);
    return result;
}

File: week09/20/pr5_Kth_smallest_Element_in_a_BST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from collections import deque
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        #Traversal - In-Order?
        visited = []
        s = deque([root])
        while s:
            curr = s[-1]
            if curr.left and curr.left.val not in visited :
                s.append(curr.left)
                continue
            else :
                visited.append(curr.val)
                s.pop()
                if len(visited) == k : return visited[-1]
                if curr.right:
                    s.append(curr.right)
        

